mPower2/mPower2/ExternalIDRegistrationViewController.swift

```diff
@@ -0,0 +1,128 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import UIKit
+import ResearchUI
+import Research
+import BridgeSDK
+
+class ExternalIDRegistrationViewController: RSDTableStepViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        // Do any additional setup after loading the view.
+    }
+
+    override func didReceiveMemoryWarning() {
+        super.didReceiveMemoryWarning()
+        // Dispose of any resources that can be recreated.
+    }
+    
+    func credentials() -> (externalID: String?, preconsent: Bool?) {
+        let resultStepIdentifier = "enterExternalId"
+        let taskPath = self.taskController.taskPath!
+        let sResult = taskPath.result.stepHistory.first { $0.identifier == resultStepIdentifier}
+        guard let stepResult = sResult as? RSDCollectionResult
+            else {
+                return (nil, nil)
+        }
+        
+        let externalIdResultIdentifier = "externalId"
+        let resultForExternalId = stepResult.inputResults.first { $0.identifier == externalIdResultIdentifier }
+        guard let eResult = resultForExternalId as? RSDAnswerResult,
+            let externalId = eResult.value as? String
+            else {
+                return (nil, nil)
+        }
+        let preConsentResultIdentifier = "preConsent"
+        let resultForPreConsent = stepResult.inputResults.first { $0.identifier == preConsentResultIdentifier }
+        guard let pResult = resultForPreConsent as? RSDAnswerResult,
+            let preconsent = pResult.value as? Bool
+            else {
+                return (nil, nil)
+        }
+
+        return (externalId, preconsent)
+    }
+
+    
+    func signUpAndSignIn(completion: @escaping SBBNetworkManagerCompletionBlock) {
+        let (ex, pc) = self.credentials()
+        guard let externalId = ex, let preconsent = pc, externalId.isEmpty == false else {
+            return
+        }
+        
+        let signUp: SBBSignUp = SBBSignUp()
+        signUp.checkForConsent = true
+        signUp.externalId = externalId
+        signUp.password = externalId
+        if preconsent {
+            signUp.dataGroups = ["test_user", "test_no_consent"]
```

Comment:
Can't set in the researcher UI because in this case, you have to import the external IDs *without* creating credentials (an account) for them. If you create credentials the old way it's the old email/password hack; if you create them the new way there's no password available to log in with so you either have to add a phone number to SMS a sign-in link to, or use a not-quite-yet-available researcher UI to log in as a researcher *in the app* to set a secret random password on the account and log in with that initially. Instead, we'll just tell Bridge about the external ID and let the app create the account (including data groups).

I'll add the comment about using a cohort rule if/when we move this particular piece of code to BridgeApp though.

Reply:
(long pause)

... oh.